File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        prefixes = list()
        maxLength = 0;
        for c in s:
            if( len(prefixes) == 0):
                maxLength+=1
                prefixes.append(c)
            else:
                lastPrefix = prefixes[-1]
                indexToSlice = lastPrefix.rfind(c)
                prefixes.append(lastPrefix[indexToSlice+1:] + c)
                maxLength = max(len(prefixes[-1]), maxLength)
        return maxLength
        

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String c) {
        if(c.length() == 0 || c.length() == 1) {
            return c.length();
        }
        int start = 0;
        int end = 0;
        int maxLength = 0; 
        
        
        while(end < c.length()-1) {
            end++;
            int i = end;
            while( i > start && i > 0 && c.charAt(i-1) != c.charAt(end)) {
                i--;
            }
            start = i;
            maxLength = Math.max(maxLength, end-start +1);
        }
        
        
        return maxLength;
    }
}



// class Solution {
//     public int lengthOfLongestSubstring(String s) {
//         if(s.length() == 0 || s.length() == 1) {
//             return s.length();
//         }
//         int start = 0;
//         int end = 0;
//         int maxLength = 0; 
//         char[] c = s.toCharArray();
//         HashMap<Character, Integer> set = new HashMap<>();
//         while(end < c.length-1) {
//             end++;
//             if(set.containsKey(c[end])) {
//                 start = set.get(c[end]) +1;
                
//                 set.put(c[end], end);
                
//             } else 
//             start = i;
//             maxLength = Math.max(maxLength, end-start +1);
//         }
        
        
//         return maxLength;
//     }
// }